### PR TITLE
Trimmed inline-end margins for flex items in horizontal writing-mode should be reflected in computed style.

### PR DESCRIPTION
**Pull request by sammygill**
#### 176479519f3166a6f19a9b10a1efcd1123058a18
<pre>
Trimmed inline-end margins for flex items in horizontal writing-mode should be reflected in computed style.
<a href="https://bugs.webkit.org/show_bug.cgi?id=253715">https://bugs.webkit.org/show_bug.cgi?id=253715</a>
rdar://106559532

Reviewed by Alan Baradlay.

When trimming the inline-end margins in RenderFlexibleBox::trimMainAxisMarginEnd
and RenderFlexibleBox::trimCrossAxisMarginEnd, instead of trimming the
margins there directly, we can replace the trimming with calls to
RenderBlock::setTrimmedMarginForChild with MarginTrimType::InlineEnd as
an argument. This will both trim the margins by calling
setMarginEndForChild and also set the appropriate rare data bit to
indicate that the inline-end margin for this box has been trimmed. In
horizonal writing modes this bit that is set should refer to
margin-right. This can be used after layout has completed to check if a
box has a trimmed margin.

When ComputedStyleExtractor tries to obtain the &quot;margin-right,&quot; value,
it will need to check if this margin has been trimmed. This is done by
calling RenderBox::hasTrimmedMargin(PhysicalDirection::Right), which
will transform this physical direction into a flow relative direction.
In horizontal writing mode with LTR direction, the &quot;right,&quot;
physical direction should correspond to the &quot;inline-end,&quot;
FlowRelativeDirection. hasTrimmedMargin should then take this new
flow relative direction and check to see if the box has this margin
marked as trimmed. If this returns true, then this means that the
box must have had its margin-right trimmed during layout and
ComputedStyleExtractor should consule the box&apos;s m_marinBox for the
trimmed value.

* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/computed-margin-values/flexbox-column-inline-end-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/computed-margin-values/flexbox-column-inline-end.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/computed-margin-values/flexbox-column-multi-line-inline-end-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/computed-margin-values/flexbox-column-multi-line-inline-end.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/computed-margin-values/flexbox-row-inline-end-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/computed-margin-values/flexbox-row-inline-end.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/computed-margin-values/flexbox-row-multi-line-inline-end-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/computed-margin-values/flexbox-row-multi-line-inline-end.html: Added.
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::rendererContainingBlockHasMarginTrim):
(WebCore::isLayoutDependent):
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle):
* Source/WebCore/rendering/RenderBlock.cpp:
(WebCore::RenderBlock::setTrimmedMarginForChild):
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::physicalToFlowRelativeDirectionMapping const):
(WebCore::RenderBox::hasTrimmedMargin const):
* Source/WebCore/rendering/RenderFlexibleBox.cpp:
(WebCore::RenderFlexibleBox::trimMainAxisMarginEnd):
(WebCore::RenderFlexibleBox::trimCrossAxisMarginEnd):

Canonical link: <a href="https://commits.webkit.org/262663@main">https://commits.webkit.org/262663@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/07a7142c51b601e3969398a3b9385c68f2e08e14

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/2175 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/2208 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/2276 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/3099 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/2217 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/2148 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/2289 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/2254 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/1968 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/2195 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/1942 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/1952 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/2955 "Built successfully") | 
| | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/1926 "1 api test failed or timed out") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/1930 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/1825 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/2006 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/1978 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/3116 "run-api-tests-without-change (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/1994 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/1787 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/1920 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/1940 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/544 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/2111 "Built successfully") | | | 
| | | | |

### CODE REVIEW

File: Source/WebCore/css/ComputedStyleExtractor.cpp [hunk 1]
```diff
@@ -2285,6 +2286,25 @@ static inline bool isNonReplacedInline(RenderObject& renderer)
...
+static bool isFlexItem(RenderObject* renderer)
+{
+    return renderer && downcast<RenderBox>(*renderer).isFlexItem();
+}
+
+static bool rendererContainingBlockHasMarginTrim(const RenderBox& renderer, std::optional<MarginTrimType> marginTrimType)
+{
+    ASSERT(renderer.isFlexItem());
+    auto* containingBlock = renderer.containingBlock();
+    if (is<HTMLHtmlElement>(renderer.element()) || containingBlock->isRenderView())
```
Comment (shivamidow) [inline Source/WebCore/css/ComputedStyleExtractor.cpp]: /* Unofficial Review */
We may need a null check for containingBlock.

File: Source/WebCore/rendering/RenderBlock.cpp [hunk 2]
```diff
@@ -3040,6 +3040,24 @@ bool RenderBlock::updateFragmentRangeForBoxChild(const RenderBox& box) const
     return false;
 }
 
+void RenderBlock::setTrimmedMarginForChild(RenderBox &child, MarginTrimType marginTrimType)
+{
+    switch (marginTrimType) {
+    case MarginTrimType::BlockStart: {
```
Comment (shivamidow) [inline Source/WebCore/rendering/RenderBlock.cpp]: Nit: The bracket is unnecessary.

File: Source/WebCore/rendering/RenderBlock.cpp [hunk 3]
```diff
@@ -3040,6 +3040,24 @@ bool RenderBlock::updateFragmentRangeForBoxChild(const RenderBox& box) const
...
+        setMarginBeforeForChild(child, 0_lu);
+        child.markMarginAsTrimmed(MarginTrimType::BlockStart);
+        break;
+    }
+    case MarginTrimType::InlineEnd: {
```
Comment (shivamidow) [inline Source/WebCore/rendering/RenderBlock.cpp]: Ditto.

File: Source/WebCore/css/ComputedStyleExtractor.cpp [hunk 4]
```diff
@@ -2721,6 +2742,29 @@ bool ComputedStyleExtractor::hasProperty(CSSPropertyID propertyID)
...
+    if (!box)
+        return false;
+
+    switch (propertyId) {
+    case CSSPropertyMarginTop:
+        return box->isFlexItem() && box->hasTrimmedMargin(PhysicalDirection::Top);
+    case CSSPropertyMarginRight:
+        return box->isFlexItem() && box->hasTrimmedMargin(PhysicalDirection::Right);
+    case CSSPropertyMarginLeft:
+
```
Comment (shivamidow) [inline Source/WebCore/css/ComputedStyleExtractor.cpp]: /* Unofficial Review */
Intended fallthrough? If so, this new line is redundant.

File: Source/WebCore/rendering/RenderBox.cpp [hunk 5]
```diff
@@ -1394,6 +1394,78 @@ void RenderBox::clearOverridingLogicalWidthLength()
...
 }
 
+std::optional<FlowRelativeDirection> RenderBox::physicalToFlowRelativeDirectionMapping(PhysicalDirection direction, LayoutCalculationPhase layoutCalculationPhase) const
+{
+    // FIXME: Return type should be just FlowRelativeDirection once all code paths have been implemented
+    auto isHorizontalWritingMode = style().isHorizontalWritingMode();
+    auto isLeftToRightDirection = style().isLeftToRightDirection();
+    // vertical-rl and horizontal-bt writing modes
+    auto isFlippedBlocksWritingMode = style().isFlippedBlocksWritingMode();
+    if (layoutCalculationPhase == LayoutCalculationPhase::Positioning) {
```
Comment (shivamidow) [inline Source/WebCore/rendering/RenderBox.cpp]: /* Unofficial Review */
Using an early return, we can mitigate the if depth.

```
if (layoutCalculationPhaase != LayoutCalculationPhase::Positioning) {
    ASSERT_NOT_IMPLEMENTED_YET();
    return {}; // std::nullopt?
}

auto isHorizontalWritingMode...;
...

if (isHorizontalWritingMode == containingBlock()->style().isHorizontalWritingMode()) {
    switch (direction) {
...
}
```

File: Source/WebCore/rendering/RenderBox.cpp [hunk 6]
```diff
@@ -1394,6 +1394,78 @@ void RenderBox::clearOverridingLogicalWidthLength()
...
+        if (isHorizontalWritingMode == containingBlock()->style().isHorizontalWritingMode()) {
+            switch (direction) {
+            case PhysicalDirection::Top:
+                return isHorizontalWritingMode ? FlowRelativeDirection::BlockStart : (isLeftToRightDirection ? FlowRelativeDirection::InlineStart : FlowRelativeDirection::InlineEnd);
```
Comment (shivamidow) [inline Source/WebCore/rendering/RenderBox.cpp]: /* Unofficial Review */
Inner ternary operators are hard to follow. I suggest you break it down with some if statements.
```
if (isHorizontalWritingMode)
    return FlowRelativeDirection::BlockStart;
return isLeftToRightDirection ? FlowRelativeDirection::InlineStart : FlowRelativeDirection::InlineEnd;
```

File: Source/WebCore/rendering/RenderBox.cpp [hunk 7]
```diff
@@ -1394,6 +1394,78 @@ void RenderBox::clearOverridingLogicalWidthLength()
...
+            case PhysicalDirection::Right:
+                return isHorizontalWritingMode ? (isLeftToRightDirection ? FlowRelativeDirection::InlineEnd : FlowRelativeDirection::InlineStart) : (isFlippedBlocksWritingMode ? FlowRelativeDirection::BlockStart : FlowRelativeDirection::BlockEnd);
```
Comment (shivamidow) [inline Source/WebCore/rendering/RenderBox.cpp]: Ditto.

File: Source/WebCore/css/ComputedStyleExtractor.cpp (hunk 1 above)
Comment (sammygill) [reply Source/WebCore/css/ComputedStyleExtractor.cpp]: I believe this should be fixed now from changes in https://github.com/WebKit/WebKit/pull/11640

File: Source/WebCore/rendering/RenderBlock.cpp (hunk 2 above)
Comment (sammygill) [reply Source/WebCore/rendering/RenderBlock.cpp]: Removed

File: Source/WebCore/rendering/RenderBlock.cpp (hunk 3 above)
Comment (sammygill) [reply Source/WebCore/rendering/RenderBlock.cpp]: Removed

File: Source/WebCore/css/ComputedStyleExtractor.cpp (hunk 4 above)
Comment (sammygill) [reply Source/WebCore/css/ComputedStyleExtractor.cpp]: I think this was also resolved with https://github.com/WebKit/WebKit/pull/11640 but let me know if you see otherwise

File: Source/WebCore/rendering/RenderBox.cpp (hunk 5 above)
Comment (sammygill) [reply Source/WebCore/rendering/RenderBox.cpp]: We ended up not including LayoutCalculationPhase as part of https://github.com/WebKit/WebKit/pull/11640

File: Source/WebCore/rendering/RenderBox.cpp (hunk 6 above)
Comment (sammygill) [reply Source/WebCore/rendering/RenderBox.cpp]: Changed

File: Source/WebCore/rendering/RenderBox.cpp (hunk 7 above)
Comment (sammygill) [reply Source/WebCore/rendering/RenderBox.cpp]: Changed

File: Source/WebCore/rendering/RenderBox.cpp [hunk 8]
```diff
@@ -1398,11 +1398,19 @@ FlowRelativeDirection RenderBox::physicalToFlowRelativeDirectionMapping(Physical
 {
     auto isHorizontalWritingMode = style().isHorizontalWritingMode();
     auto isLeftToRightDirection = style().isLeftToRightDirection();
+    // vertical-rl and horizontal-bt writing modes
+    auto isFlippedBlocksWritingMode = style().isFlippedBlocksWritingMode();
```
Comment (alanbaradlay) [inline Source/WebCore/rendering/RenderBox.cpp]: this should be grabbed off of the ancestor (flex container, containing block etc) which sets up the writing mode for this box, see https://github.com/WebKit/WebKit/pull/11640#discussion_r1146464625

Comment (sammygill) [reply Source/WebCore/rendering/RenderBox.cpp]: I went ahead and added a lambda to get the style off of the formatting context. Currently it can grab it off of flexboxes and will `ASSERT_NOT_IMPLEMENTED_YET` for any other contexts. At the call site where `physicalToFlowRelativeDirection` is being called I added another lambda that checks to see if it can use this method so it doesn't trigger the assert

File: Source/WebCore/css/ComputedStyleExtractor.cpp [hunk 9]
```diff
@@ -2341,7 +2341,7 @@ static bool isLayoutDependent(CSSPropertyID propertyID, const RenderStyle* style
     case CSSPropertyMarginTop:
         return paddingOrMarginIsRendererDependent<&RenderStyle::marginTop>(style, renderer) || (isFlexItem(renderer) && rendererContainingBlockHasMarginTrim(downcast<RenderBox>(*renderer), MarginTrimType::BlockStart));
     case CSSPropertyMarginRight:
-        return paddingOrMarginIsRendererDependent<&RenderStyle::marginRight>(style, renderer);
+        return paddingOrMarginIsRendererDependent<&RenderStyle::marginRight>(style, renderer) || (isFlexItem(renderer) && rendererContainingBlockHasMarginTrim(downcast<RenderBox>(*renderer), MarginTrimType::InlineEnd));
```
Comment (alanbaradlay) [inline Source/WebCore/css/ComputedStyleExtractor.cpp:2350]: Could you, in a separate patch, rename rendererContainingBlockHasMarginTrim to something that's not specific to BFC, please?

File: Source/WebCore/rendering/RenderBox.cpp [hunk 10]
```diff
@@ -1396,13 +1396,28 @@ void RenderBox::clearOverridingLogicalWidthLength()
...
 FlowRelativeDirection RenderBox::physicalToFlowRelativeDirectionMapping(PhysicalDirection direction) const
 {
-    auto isHorizontalWritingMode = style().isHorizontalWritingMode();
-    auto isLeftToRightDirection = style().isLeftToRightDirection();
+    auto determineFormattingContextStyle = [&]() -> const RenderStyle& {
+        if (isFlexItem())
+            return parent()->style();
+        ASSERT_NOT_IMPLEMENTED_YET();
+        return style();
+    };
```
Comment (alanbaradlay) [inline Source/WebCore/rendering/RenderBox.cpp]: formatting contexts do not have styles (formatting context roots do), FCs have directionalities (inline, block) among other things, but you can't say e.g. this FC has the background color of green.

File: Source/WebCore/rendering/RenderBox.cpp [hunk 11]
```diff
@@ -1466,7 +1481,10 @@ bool RenderBox::hasTrimmedMargin(PhysicalDirection physicalDirection) const
 #endif
     ASSERT(!needsLayout());
 
-    if (physicalDirection == PhysicalDirection::Top)
+    auto canUsePhysicalToFlowRelativeMapping = [&]() {
+        return isFlexItem();
+    };
+    if ((physicalDirection == PhysicalDirection::Top || physicalDirection == PhysicalDirection::Right) && canUsePhysicalToFlowRelativeMapping())
```
Comment (alanbaradlay) [inline Source/WebCore/rendering/RenderBox.cpp]: so this is gonna assert anyway if called with unsupported content?

File: Source/WebCore/css/ComputedStyleExtractor.cpp (hunk 9 above)
Comment (sammygill) [reply Source/WebCore/css/ComputedStyleExtractor.cpp:2350]: Yes, I think https://github.com/WebKit/WebKit/pull/12273 could be a good patch to do it in

File: Source/WebCore/rendering/RenderBox.cpp (hunk 10 above)
Comment (sammygill) [reply Source/WebCore/rendering/RenderBox.cpp]: I changed the naming here to reference the FC root instead since that is the intention of the helper. I believe this should be the correct style to use when determining the `isFlippedBlocksWritingMode` and `isLeftToRightDirection` below, but let me know if you think that's not right

File: Source/WebCore/rendering/RenderBox.cpp (hunk 11 above)
Comment (sammygill) [reply Source/WebCore/rendering/RenderBox.cpp]: Yeah that's not quite right, thanks for catching. After taking another look I think this can be removed since any asserts should be guarded by the logic of `rendererContainingBlockhasMarginTrim` (or whatever the name of that function will become) before calling `hasTrimmedMargin`

File: Source/WebCore/rendering/RenderBox.cpp [hunk 12]
```diff
@@ -1391,13 +1391,28 @@ void RenderBox::clearOverridingLogicalWidthLength()
 
 FlowRelativeDirection RenderBox::physicalToFlowRelativeDirectionMapping(PhysicalDirection direction) const
 {
-    auto isHorizontalWritingMode = style().isHorizontalWritingMode();
-    auto isLeftToRightDirection = style().isLeftToRightDirection();
+    auto determineFormattingContextRootStyle = [&]() -> const RenderStyle& {
```
Comment (alanbaradlay) [inline Source/WebCore/rendering/RenderBox.cpp:1394]: auto determineFormattingContextRootStyle = [&] { ?

File: Source/WebCore/rendering/RenderBox.cpp [hunk 13]
```diff
@@ -1391,13 +1391,28 @@ void RenderBox::clearOverridingLogicalWidthLength()
...
+        if (isFlexItem())
+            return parent()->style();
+        ASSERT_NOT_IMPLEMENTED_YET();
+        return style();
+    };
+    const auto& formattingContextRootStyle = determineFormattingContextRootStyle();
```
Comment (alanbaradlay) [inline Source/WebCore/rendering/RenderBox.cpp]: const is redundant here.

**Review by alanbaradlay — APPROVED** [review]

File: Source/WebCore/rendering/RenderBox.cpp (hunk 12 above)
Comment (sammygill) [reply Source/WebCore/rendering/RenderBox.cpp:1394]: Unfortunately the type deduction throws an error here and deduces just `RenderStyle`. Maybe there's another way to resolve that issue but I'm not super familiar with it.
